agroture/micropython
# esp/espnow: Support for the Espressif ESP-NOW protocol.

- Micropython v1.13 - v1.19 compatible
- Ring buffers in shared/runtime/ringbuffer.[ch] for robust IO.
- Signal strength (RSSI) monitoring (see ESPNow.peers)
- Asyncio support via aioespnow.py module.
- Docs provided at docs/library/espnow.rst.
- PR6515.

Methods in espnow.ESPNow class:
- active(True/False)
- config(): set rx buffer size, read timeout and tx rate.
- recv()/irecv() to read incoming messages from peers.
- send() to send messages to peer devices
- any() to test if a message is ready to read.
- irq()/on_recv() to set callback for received messages.
- stats(): Returns transfer stats: (tx_pkts, tx_pkt_responses, tx_failures, rx_pkts, lost_rx_pkts).
- add_peer(mac, ...): register peer before sending messages.
- get_peer(mac): Return peer info: (mac,lmk,channel,ifidx,encrypt)
- mod_peer(mac, ...) to change peer info parameters.
- get_peers(): to return all peer info tuples.
- peers_table: to support RSSI signal monitoring for received messages: {peer1: [rssi, time_ms], peer2: [rssi, time_ms], ...}

aioespnow.AIOESPNow class provides coroutines:
- airecv(), arecv() and asend().

ESP8266 is a pared down version of the esp32 espnow support due to code size restrictions and differences in the low-level API. See docs for details.

Add constants as attributes of the espnow module:
- espnow.MAX_DATA_LEN         (=ESP_NOW_MAX_DATA_LEN) (250)
- espnow.KEY_LEN              (=ESP_NOW_KEY_LEN) (16)
- espnow.MAX_TOTAL_PEER_NUM   (=ESP_NOW_MAX_TOTAL_PEER_NUM) (20)
- espnow.MAX_ENCRYPT_PEER_NUM (=ESP_NOW_MAX_ENCRYPT_PEER_NUM) (6)
- espnow.ESP_NOW_ETH_ALEN     (=ESP_NOW_ETH_ALEN) (6)

Test suite in tests/multi_espnow:
- Tests basic espnow data transfer, multiple transfers various message sizes, encrypted messages (pmk and lmk), asyncio support.

Initial PR at:
https://github.com/micropython/micropython/pull/4115. Initial import of code from:
https://github.com/nickzoic/micropython/tree/espnow-4115. Including contributions from @nickzoic @shawwwn and @zoland.

esp32/8266: WLAN(): add support for set/get wifi power saving mode.

For esp32 and esp8266: adds:
- 'pm' option to WLAN.config() to set/get the wifi power saving mode; and
- PM_NONE, PM_MIN_MODEM, PM_MAX_MODEM and PM_LIGHT_SLEEP (eps8266 only) constants to the WLAN class.

espnow: Reduce code size on ESP8266.

Minor code reduction on ESP8266 to accommodate recent upstream changes:
- Delete buffer conistency check in espnow_recvinto().
- espnow_send(): only check for ROM addresses higher than GC pool.

espnow: Docs minor update.

- ESPNow.config(rate=XX) is only available on ESP32 and IDF>=4.3.0; and
- Add pointer to the espnow Issues page.
- Add note on automatic channel assignment on routers.

Use MP_DEFINE_CONST_OBJ_TYPE() to initialise esp_espnow_type.

Extend py/ringbuf.c and use that instead of my own implementation.
